pageserver/src/import_datadir.rs

```diff
@@ -617,3 +619,27 @@ async fn read_all_bytes(reader: &mut (impl AsyncRead + Unpin)) -> Result<Bytes>
     reader.read_to_end(&mut buf).await?;
     Ok(Bytes::from(buf))
 }
+
+pub async fn create_tar_zst(pgdata_path: &Path) -> Result<Vec<u8>> {
+    let mut paths = Vec::new();
+    for entry in WalkDir::new(pgdata_path) {
+        let entry = entry?;
+        let metadata = entry.metadata().expect("error getting dir entry metadata");
+        if metadata.is_file() {
+            paths.push(entry.path().to_owned());
+        }
+    }
+    // Don't rely on file system order for listing as it may be non-deterministic
+    paths.sort();
+    let zstd = async_compression::tokio::write::ZstdEncoder::new(Vec::new());
+    let mut builder = Builder::new(zstd);
+    // Use reproducible header mode
+    builder.mode(HeaderMode::Deterministic);
+    for path in paths {
+        builder.append_path(path).await?;
+    }
+    builder.finish().await?;
+    let zstd = builder.into_inner().await?;
+    let compressed = zstd.into_inner();
+    Ok(compressed)
```

Comment:
I cannot remember looking how often does async_compression yield, but this is a good example of something which would be great as spawn_blocking.

Reply:
spawn_blocking doesn't accept async code. So we'd have to drop async-compression and tokio-tar entirely, and use zstd like in the earlier commit. I'm not a fan of that.

Reply:
ideally, async-compression would yield each time after another block has been compressed and shouldn't rely on the underlying I/O structure to do the yielding. Maybe we could just impl `AsyncWrite` manually on a new yielding vec wrapper or something which yields every >= 1 kb of data?